File: src/lib/emi-calculator.ts
/**
 * EMI (Equated Monthly Installment) calculator per PRD §9.2.
 *
 * EMI = P × r × (1+r)^n / ((1+r)^n - 1)
 *   where r = annual_interest_rate / 12 / 100 and n = tenure in months.
 *   When r is 0, the formula degenerates to EMI = P / n.
 *
 * All return values are strings with exactly 2 decimal places so the caller
 * can hand them directly to Prisma's Decimal columns. The floating-point
 * calculation is contained inside this module — never propagate JS numbers
 * for money past this boundary.
 *
 * Rounding strategy: each month's interest and principal are rounded to 2
 * decimals independently, and the FINAL payment absorbs any residual balance
 * so remainingBalance lands on exactly 0.00. This matches how banks present
 * amortization schedules and avoids -0.01 / +0.03 drift at the end.
 */

export interface CalculateEmiInput {
  principal: string | number
  annualRate: string | number // percent, e.g. 12 for 12%
  tenureMonths: number
}

export interface EmiBreakdown {
  /** Rounded monthly EMI, 2 decimals. */
  emiAmount: string
  /** Total paid over the full schedule, 2 decimals. */
  totalPayment: string
  /** Total interest paid = totalPayment - principal, 2 decimals. */
  totalInterest: string
}

export interface AmortizationRow {
  paymentNumber: number
  dueDate: Date
  emiAmount: string
  principalComponent: string
  interestComponent: string
  remainingBalance: string
}

function toNumber(value: string | number): number {
  return typeof value === 'number' ? value : Number(value)
}

function round2(value: number): number {
  return Math.round(value * 100) / 100
}

/**
 * Compute the monthly EMI, total payment, and total interest for a loan.
 * Useful for the Add EMI form's live preview card.
 */
export function calculateEmi(input: CalculateEmiInput): EmiBreakdown {
  const principal = toNumber(input.principal)
  const annualRate = toNumber(input.annualRate)
  const n = input.tenureMonths

  if (!Number.isFinite(principal) || principal <= 0) {
    throw new Error('Principal must be greater than 0')
  }
  if (!Number.isFinite(annualRate) || annualRate < 0) {
    throw new Error('Annual rate must be 0 or greater')
  }
  if (!Number.isFinite(n) || n <= 0 || !Number.isInteger(n)) {
    throw new Error('Tenure must be a positive integer')
  }

  const r = annualRate / 12 / 100

  let emi: number
  if (r === 0) {
    emi = principal / n
  } else {
    const pow = Math.pow(1 + r, n)
    emi = (principal * r * pow) / (pow - 1)
  }

  const emiRounded = round2(emi)
  const totalPayment = round2(emiRounded * n)
  const totalInterest = round2(totalPayment - principal)

  return {
    emiAmount: emiRounded.toFixed(2),
    totalPayment: totalPayment.toFixed(2),
    totalInterest: totalInterest.toFixed(2),
  }
}

/**
 * Add `months` calendar months to a date, preserving day-of-month when
 * possible and clamping to the end of the target month otherwise
 * (e.g. Jan 31 + 1 month → Feb 28/29). Operates in UTC so amortization
 * due dates are identical across server and client regardless of timezone.
 */
function addMonths(date: Date, months: number): Date {
  const year = date.getUTCFullYear()
  const month = date.getUTCMonth()
  const day = date.getUTCDate()
  const targetDate = new Date(Date.UTC(year, month + months, day))
  // If JS rolled the date forward (e.g. Jan 31 → Mar 3), clamp to last day
  // of the intended month.
  const expectedMonth = (((month + months) % 12) + 12) % 12
  if (targetDate.getUTCMonth() !== expectedMonth) {
    return new Date(Date.UTC(year, month + months + 1, 0))
  }
  return targetDate
}

/**
 * Generate the full amortization schedule.
 *
 * Returns one row per month. The last row's principal component is forced to
 * consume whatever principal is still outstanding so remainingBalance === 0.00
 * after the final payment.
 */
export function generateAmortization(input: {
  principal: string | number
  annualRate: string | number
  tenureMonths: number
  startDate: Date
}): AmortizationRow[] {
  const principal = toNumber(input.principal)
  const annualRate = toNumber(input.annualRate)
  const n = input.tenureMonths
  const { emiAmount } = calculateEmi({
    principal,
    annualRate,
    tenureMonths: n,
  })
  const emi = Number(emiAmount)
  const r = annualRate / 12 / 100

  const rows: AmortizationRow[] = []
  let balance = principal

  for (let i = 1; i <= n; i++) {
    const isLast = i === n

    const interestComponent = round2(balance * r)
    let principalComponent: number
    let paymentAmount: number

    if (isLast) {
      // Force the final payment to clear the balance exactly.
      principalComponent = round2(balance)
      paymentAmount = round2(principalComponent + interestComponent)
      balance = 0
    } else {
      principalComponent = round2(emi - interestComponent)
      // Guard against tiny negative principal if interest > emi on a weird
      // rate/tenure combo (shouldn't happen in normal use, but be safe).
      if (principalComponent < 0) principalComponent = 0
      paymentAmount = emi
      balance = round2(balance - principalComponent)
    }

    rows.push({
      paymentNumber: i,
      dueDate: addMonths(input.startDate, i - 1),
      emiAmount: paymentAmount.toFixed(2),
      principalComponent: principalComponent.toFixed(2),
      interestComponent: interestComponent.toFixed(2),
      remainingBalance: balance.toFixed(2),
    })
  }

  return rows
}

File: apps/ios/Phinio/Phinio/Domain/EmiCalculator.swift
import Foundation

/// Port of src/lib/emi-calculator.ts (PRD §9.2). Deliberately computes in
/// Double with the same operations in the same order as the TS original so
/// offline-created schedules match the server to the paisa — proven by
/// PhinioTests/Fixtures/emi-fixtures.json. Do not "upgrade" this to Decimal.
enum EmiMethod: String {
  case bankLoan = "bank_loan"
  case creditCard = "credit_card"
}

struct EmiBreakdown {
  let emiAmount: String
  let totalPayment: String
  let totalInterest: String
}

struct AmortizationRow {
  let paymentNumber: Int
  let dueDate: Date
  let emiAmount: String
  let principalComponent: String
  let interestComponent: String
  let remainingBalance: String
}

enum EmiCalculatorError: Error {
  case invalidPrincipal, invalidRate, invalidTenure
}

enum EmiCalculator {
  static let feePaymentNumber = 0

  /// Matches JS `Math.round(value * 100) / 100` (half-up for positives).
  private static func round2(_ value: Double) -> Double {
    (value * 100).rounded(.toNearestOrAwayFromZero) / 100
  }

  /// Matches JS `x.toFixed(2)` for the already-round2'ed values we emit.
  private static func fixed2(_ value: Double) -> String {
    let cents = Int((value * 100).rounded(.toNearestOrAwayFromZero))
    return String(format: "%d.%02d", cents / 100, abs(cents % 100))
  }

  static func calculate(
    principal: Double, annualRate: Double, tenureMonths: Int, type: EmiMethod
  ) throws -> EmiBreakdown {
    guard principal.isFinite, principal > 0 else { throw EmiCalculatorError.invalidPrincipal }
    guard annualRate.isFinite, annualRate >= 0 else { throw EmiCalculatorError.invalidRate }
    guard tenureMonths > 0 else { throw EmiCalculatorError.invalidTenure }

    let n = Double(tenureMonths)
    let r = annualRate / 12 / 100

    let emi: Double
    if r == 0 {
      emi = principal / n
    } else if type == .creditCard {
      emi = principal / n + principal * r
    } else {
      let pow = Foundation.pow(1 + r, n)
      emi = (principal * r * pow) / (pow - 1)
    }

    let emiRounded = round2(emi)
    let totalPayment = round2(emiRounded * n)
    let totalInterest = round2(totalPayment - principal)
    return EmiBreakdown(
      emiAmount: fixed2(emiRounded),
      totalPayment: fixed2(totalPayment),
      totalInterest: fixed2(totalInterest))
  }

  /// Add calendar months in UTC, clamping to end-of-month like the TS
  /// original (Jan 31 + 1 → Feb 28/29).
  private static func addMonths(_ date: Date, _ months: Int) -> Date {
    var cal = Calendar(identifier: .gregorian)
    cal.timeZone = TimeZone(identifier: "UTC")!
    let parts = cal.dateComponents([.year, .month, .day], from: date)
    let year = parts.year!, month = parts.month!, day = parts.day!

    let zeroBased = (month - 1) + months
    let targetYear = year + Int(floor(Double(zeroBased) / 12))
    let targetMonth = ((zeroBased % 12) + 12) % 12 + 1

    var target = DateComponents(year: targetYear, month: targetMonth, day: 1)
    let firstOfMonth = cal.date(from: target)!
    let daysInMonth = cal.range(of: .day, in: .month, for: firstOfMonth)!.count
    target.day = min(day, daysInMonth)
    return cal.date(from: target)!
  }

  static func amortization(
    principal: Double, annualRate: Double, tenureMonths: Int,
    startDate: Date, type: EmiMethod
  ) throws -> [AmortizationRow] {
    let breakdown = try calculate(
      principal: principal, annualRate: annualRate,
      tenureMonths: tenureMonths, type: type)
    let emi = Double(breakdown.emiAmount)!
    let r = annualRate / 12 / 100
    let isFlat = type == .creditCard

    var rows: [AmortizationRow] = []
    var balance = principal

    for i in 1...tenureMonths {
      let isLast = i == tenureMonths
      let interestComponent = round2(isFlat ? principal * r : balance * r)
      let principalComponent: Double
      let paymentAmount: Double

      if isLast {
        principalComponent = round2(balance)
        paymentAmount = round2(principalComponent + interestComponent)
        balance = 0
      } else {
        principalComponent = max(0, round2(emi - interestComponent))
        paymentAmount = emi
        balance = round2(balance - principalComponent)
      }

      rows.append(AmortizationRow(
        paymentNumber: i,
        dueDate: addMonths(startDate, i - 1),
        emiAmount: fixed2(paymentAmount),
        principalComponent: fixed2(principalComponent),
        interestComponent: fixed2(interestComponent),
        remainingBalance: fixed2(balance)))
    }
    return rows
  }
}
